File: Zig/progress_bar/progress_bar.zig
const std = @import("std");

/// Progress bar configuration options
pub const ProgressBarOptions = struct {
    width: usize = 40,
    fill_char: u8 = '=',
    empty_char: u8 = '-',
    left_bracket: u8 = '[',
    right_bracket: u8 = ']',
    show_percentage: bool = true,
    show_count: bool = false,
    show_time: bool = false,
    show_eta: bool = false,
    prefix: []const u8 = "",
    suffix: []const u8 = "",
};

/// A terminal progress bar
pub const ProgressBar = struct {
    allocator: std.mem.Allocator,
    total: usize,
    current: usize,
    options: ProgressBarOptions,
    start_time: ?i64,
    last_update: i64,
    update_interval_ms: i64,
    writer: std.io.AnyWriter,

    const Self = @This();

    /// Initialize a new progress bar
    pub fn init(allocator: std.mem.Allocator, total: usize, writer: std.io.AnyWriter, options: ProgressBarOptions) Self {
        return .{
            .allocator = allocator,
            .total = total,
            .current = 0,
            .options = options,
            .start_time = null,
            .last_update = 0,
            .update_interval_ms = 100,
            .writer = writer,
        };
    }

    /// Start the progress bar
    pub fn start(self: *Self) !void {
        self.start_time = std.time.milliTimestamp();
        self.last_update = 0;
        try self.render();
    }

    /// Update progress by increment
    pub fn increment(self: *Self) !void {
        try self.update(self.current + 1);
    }

    /// Update progress to a specific value
    pub fn update(self: *Self, current: usize) !void {
        self.current = @min(current, self.total);
        
        const now = std.time.milliTimestamp();
        if (now - self.last_update < self.update_interval_ms and self.current < self.total) {
            return;
        }
        self.last_update = now;
        
        try self.render();
    }

    /// Finish the progress bar
    pub fn finish(self: *Self) !void {
        self.current = self.total;
        try self.render();
        try self.writer.print("\n", .{});
    }

    /// Get current percentage (0-100)
    pub fn percentage(self: Self) usize {
        if (self.total == 0) return 100;
        return @as(usize, @intCast(@divTrunc(@as(i64, @intCast(self.current)) * 100, @as(i64, @intCast(self.total)))));
    }

    /// Get elapsed time in milliseconds
    pub fn elapsedMs(self: Self) i64 {
        if (self.start_time == null) return 0;
        return std.time.milliTimestamp() - self.start_time.?;
    }

    /// Get estimated remaining time in milliseconds
    pub fn etaMs(self: Self) i64 {
        if (self.current == 0 or self.start_time == null) return 0;
        const elapsed = self.elapsedMs();
        if (elapsed == 0) return 0;
        
        const rate = @as(f64, @floatFromInt(self.current)) / @as(f64, @floatFromInt(elapsed));
        const remaining = @as(f64, @floatFromInt(self.total - self.current));
        
        return @as(i64, @intFromFloat(remaining / rate));
    }

    /// Format duration as human-readable string
    fn formatDuration(ms: i64, writer: std.io.AnyWriter) !void {
        const total_seconds = @divTrunc(ms, 1000);
        const hours = @divTrunc(total_seconds, 3600);
        const minutes = @divTrunc(@mod(total_seconds, 3600), 60);
        const seconds = @mod(total_seconds, 60);
        
        if (hours > 0) {
            try writer.print("{d:0>2}:{d:0>2}:{d:0>2}", .{ hours, minutes, seconds });
        } else if (minutes > 0) {
            try writer.print("{d:0>2}:{d:0>2}", .{ minutes, seconds });
        } else {
            try writer.print("{d}s", .{seconds});
        }
    }

    /// Render the progress bar
    fn render(self: Self) !void {
        // Clear current line and move to beginning
        try self.writer.print("\x1b[2K\r", .{});

        // Print prefix
        if (self.options.prefix.len > 0) {
            try self.writer.print("{s} ", .{self.options.prefix});
        }

        // Print left bracket
        try self.writer.writeByte(self.options.left_bracket);

        // Calculate filled portion
        const pct = self.percentage();
        const filled = @as(usize, @intFromFloat(@as(f64, @floatFromInt(pct)) / 100.0 * @as(f64, @floatFromInt(self.options.width))));

        // Print filled portion
        var i: usize = 0;
        while (i < filled) : (i += 1) {
            try self.writer.writeByte(self.options.fill_char);
        }

        // Print empty portion
        while (i < self.options.width) : (i += 1) {
            try self.writer.writeByte(self.options.empty_char);
        }

        // Print right bracket
        try self.writer.writeByte(self.options.right_bracket);

        // Print percentage
        if (self.options.show_percentage) {
            try self.writer.print(" {d:3}%", .{pct});
        }

        // Print count
        if (self.options.show_count) {
            try self.writer.print(" ({d}/{d})", .{ self.current, self.total });
        }

        // Print elapsed time
        if (self.options.show_time) {
            try self.writer.print(" [", .{});
            try formatDuration(self.elapsedMs(), self.writer);
            try self.writer.writeByte(']');
        }

        // Print ETA
        if (self.options.show_eta and self.current > 0) {
            try self.writer.print(" ETA: ", .{});
            try formatDuration(self.etaMs(), self.writer);
        }

        // Print suffix
        if (self.options.suffix.len > 0) {
            try self.writer.print(" {s}", .{self.options.suffix});
        }
    }
};

/// Multi-progress bar manager for tracking multiple items
pub const MultiProgressBar = struct {
    allocator: std.mem.Allocator,
    bars: std.ArrayList(*ProgressBar),
    writer: std.io.AnyWriter,
    last_render_count: usize,

    const Self = @This();

    /// Initialize multi-progress bar
    pub fn init(allocator: std.mem.Allocator, writer: std.io.AnyWriter) Self {
        return .{
            .allocator = allocator,
            .bars = std.ArrayList(*ProgressBar).init(allocator),
            .writer = writer,
            .last_render_count = 0,
        };
    }

    /// Deinitialize
    pub fn deinit(self: *Self) void {
        self.bars.deinit();
    }

    /// Add a progress bar
    pub fn addBar(self: *Self, bar: *ProgressBar) !void {
        try self.bars.append(bar);
        bar.update_interval_ms = 0; // We'll control rendering
    }

    /// Render all bars
    pub fn render(self: *Self) !void {
        // Move cursor up for previous bars
        if (self.last_render_count > 0) {
            try self.writer.print("\x1b[{d}A", .{self.last_render_count});
        }

        for (self.bars.items) |bar| {
            try bar.render();
            try self.writer.print("\n", .{});
        }

        self.last_render_count = self.bars.items.len;
    }
};

/// Spinner for indeterminate progress
pub const Spinner = struct {
    frames: []const u8,
    current_frame: usize,
    interval_ms: i64,
    last_update: i64,
    writer: std.io.AnyWriter,
    message: []const u8,
    done: bool,

    const Self = @This();

    /// Default spinner frames
    pub const default_frames = "|/-\\";

    /// Initialize spinner
    pub fn init(writer: std.io.AnyWriter, message: []const u8) Self {
        return .{
            .frames = default_frames,
            .current_frame = 0,
            .interval_ms = 100,
            .last_update = 0,
            .writer = writer,
            .message = message,
            .done = false,
        };
    }

    /// Initialize spinner with custom frames
    pub fn initWithFrames(writer: std.io.AnyWriter, message: []const u8, frames: []const u8) Self {
        return .{
            .frames = frames,
            .current_frame = 0,
            .interval_ms = 100,
            .last_update = 0,
            .writer = writer,
            .message = message,
            .done = false,
        };
    }

    /// Update spinner
    pub fn tick(self: *Self) !void {
        if (self.done) return;

        const now = std.time.milliTimestamp();
        if (now - self.last_update < self.interval_ms) return;
        self.last_update = now;

        try self.writer.print("\r\x1b[2K{s} {s}", .{ self.frames[self.current_frame..][0..1], self.message });
        
        self.current_frame = (self.current_frame + 1) % self.frames.len;
    }

    /// Complete spinner
    pub fn complete(self: *Self, final_message: []const u8) !void {
        self.done = true;
        try self.writer.print("\r\x1b[2K{s}\n", .{final_message});
    }
};

// ============================================================================
// Tests
// ============================================================================

test "ProgressBar initialization" {
    var buffer: [1024]u8 = undefined;
    var stream = std.io.fixedBufferStream(&buffer);
    const writer = stream.writer().any();
    
    const bar = ProgressBar.init(std.testing.allocator, 100, writer, .{});
    try std.testing.expectEqual(@as(usize, 100), bar.total);
    try std.testing.expectEqual(@as(usize, 0), bar.current);
    try std.testing.expectEqual(@as(usize, 40), bar.options.width);
}

test "ProgressBar percentage calculation" {
    var buffer: [1024]u8 = undefined;
    var stream = std.io.fixedBufferStream(&buffer);
    const writer = stream.writer().any();
    
    var bar = ProgressBar.init(std.testing.allocator, 100, writer, .{});
    try std.testing.expectEqual(@as(usize, 0), bar.percentage());
    
    bar.current = 50;
    try std.testing.expectEqual(@as(usize, 50), bar.percentage());
    
    bar.current = 100;
    try std.testing.expectEqual(@as(usize, 100), bar.percentage());
}

test "ProgressBar percentage with non-100 total" {
    var buffer: [1024]u8 = undefined;
    var stream = std.io.fixedBufferStream(&buffer);
    const writer = stream.writer().any();
    
    var bar = ProgressBar.init(std.testing.allocator, 250, writer, .{});
    _ = &bar; // force mutation
    bar.current = 125;
    try std.testing.expectEqual(@as(usize, 50), bar.percentage());
}

test "ProgressBar zero total returns 100%" {
    var buffer: [1024]u8 = undefined;
    var stream = std.io.fixedBufferStream(&buffer);
    const writer = stream.writer().any();
    
    const bar = ProgressBar.init(std.testing.allocator, 0, writer, .{});
    try std.testing.expectEqual(@as(usize, 100), bar.percentage());
}

test "ProgressBar update respects max" {
    var buffer: [1024]u8 = undefined;
    var stream = std.io.fixedBufferStream(&buffer);
    const writer = stream.writer().any();
    
    var bar = ProgressBar.init(std.testing.allocator, 100, writer, .{});
    bar.start_time = std.time.milliTimestamp();
    
    try bar.update(150);
    try std.testing.expectEqual(@as(usize, 100), bar.current);
}

test "ProgressBar custom options" {
    var buffer: [1024]u8 = undefined;
    var stream = std.io.fixedBufferStream(&buffer);
    const writer = stream.writer().any();
    
    const bar = ProgressBar.init(std.testing.allocator, 100, writer, .{
        .width = 20,
        .fill_char = '#',
        .empty_char = '.',
        .prefix = "Processing",
        .show_count = true,
    });
    
    try std.testing.expectEqual(@as(usize, 20), bar.options.width);
    try std.testing.expectEqual(@as(u8, '#'), bar.options.fill_char);
    try std.testing.expectEqual(@as(u8, '.'), bar.options.empty_char);
}

test "Spinner initialization" {
    var buffer: [1024]u8 = undefined;
    var stream = std.io.fixedBufferStream(&buffer);
    const writer = stream.writer().any();
    
    const spinner = Spinner.init(writer, "Loading...");
    try std.testing.expectEqualStrings("|/-\\", spinner.frames);
    try std.testing.expectEqual(@as(usize, 0), spinner.current_frame);
    try std.testing.expectEqualStrings("Loading...", spinner.message);
}

test "Spinner custom frames" {
    var buffer: [1024]u8 = undefined;
    var stream = std.io.fixedBufferStream(&buffer);
    const writer = stream.writer().any();
    
    const spinner = Spinner.initWithFrames(writer, "Working", "⣾⣽⣻⢿⡿⣟⣯⣷");
    try std.testing.expectEqualStrings("⣾⣽⣻⢿⡿⣟⣯⣷", spinner.frames);
    try std.testing.expectEqualStrings("Working", spinner.message);
}

test "ProgressBar elapsed time" {
    var buffer: [1024]u8 = undefined;
    var stream = std.io.fixedBufferStream(&buffer);
    const writer = stream.writer().any();
    
    var bar = ProgressBar.init(std.testing.allocator, 100, writer, .{});
    _ = &bar; // force mutation
    try std.testing.expectEqual(@as(i64, 0), bar.elapsedMs());
    
    bar.start_time = std.time.milliTimestamp() - 5000;
    const elapsed = bar.elapsedMs();
    try std.testing.expect(elapsed >= 5000);
}

test "ProgressBar ETA calculation" {
    var buffer: [1024]u8 = undefined;
    var stream = std.io.fixedBufferStream(&buffer);
    const writer = stream.writer().any();
    
    var bar = ProgressBar.init(std.testing.allocator, 100, writer, .{});
    
    // No progress, no ETA
    try std.testing.expectEqual(@as(i64, 0), bar.etaMs());
    
    // Half done after 1 second = 1 second remaining
    bar.start_time = std.time.milliTimestamp() - 1000;
    bar.current = 50;
    const eta = bar.etaMs();
    // Should be approximately 1 second (1000ms), allow some tolerance
    try std.testing.expect(eta > 500 and eta < 2000);
}

test "MultiProgressBar initialization" {
    const allocator = std.testing.allocator;
    var buffer: [2048]u8 = undefined;
    var stream = std.io.fixedBufferStream(&buffer);
    const writer = stream.writer().any();
    
    var multi = MultiProgressBar.init(allocator, writer);
    defer multi.deinit();
    
    try std.testing.expectEqual(@as(usize, 0), multi.bars.items.len);
}

test "MultiProgressBar addBar" {
    const allocator = std.testing.allocator;
    var buffer: [2048]u8 = undefined;
    var stream = std.io.fixedBufferStream(&buffer);
    const writer = stream.writer().any();
    
    var multi = MultiProgressBar.init(allocator, writer);
    defer multi.deinit();
    
    var bar1 = ProgressBar.init(allocator, 100, writer, .{});
    var bar2 = ProgressBar.init(allocator, 50, writer, .{});
    
    try multi.addBar(&bar1);
    try multi.addBar(&bar2);
    
    try std.testing.expectEqual(@as(usize, 2), multi.bars.items.len);
}

File: Go/progress_bar/progress_bar.go
// Package progress_bar provides a terminal progress bar utility with zero external dependencies.
// Supports multiple styles, colors, ETA estimation, and concurrent safety.
package progress_bar

import (
	"fmt"
	"io"
	"os"
	"strings"
	"sync"
	"time"
)

// Style represents a progress bar visual style
type Style struct {
	Complete    string // Character for completed portion
	Incomplete  string // Character for incomplete portion
	LeftEnd     string // Left border character
	RightEnd    string // Right border character
	Description string // Style description
}

// Predefined progress bar styles
var (
	StyleDefault = Style{
		Complete:    "█",
		Incomplete:  "░",
		LeftEnd:     "|",
		RightEnd:    "|",
		Description: "Default block style",
	}

	StyleClassic = Style{
		Complete:    "=",
		Incomplete:  " ",
		LeftEnd:     "[",
		RightEnd:    "]",
		Description: "Classic ASCII style",
	}

	StyleArrow = Style{
		Complete:    "=",
		Incomplete:  " ",
		LeftEnd:     "[",
		RightEnd:    ">",
		Description: "Arrow style",
	}

	StyleBlocks = Style{
		Complete:    "▓",
		Incomplete:  "░",
		LeftEnd:     "",
		RightEnd:    "",
		Description: "Block style without borders",
	}

	StyleDots = Style{
		Complete:    "●",
		Incomplete:  "○",
		LeftEnd:     "",
		RightEnd:    "",
		Description: "Dots style",
	}

	StylePipe = Style{
		Complete:    "│",
		Incomplete:  " ",
		LeftEnd:     "├",
		RightEnd:    "┤",
		Description: "Pipe style",
	}

	StyleMinimal = Style{
		Complete:    "#",
		Incomplete:  "-",
		LeftEnd:     "",
		RightEnd:    "",
		Description: "Minimal style",
	}

	StyleCircle = Style{
		Complete:    "◉",
		Incomplete:  "◎",
		LeftEnd:     "",
		RightEnd:    "",
		Description: "Circle style",
	}
)

// Color represents ANSI color codes
type Color string

const (
	ColorReset  Color = "\033[0m"
	ColorRed    Color = "\033[31m"
	ColorGreen  Color = "\033[32m"
	ColorYellow Color = "\033[33m"
	ColorBlue   Color = "\033[34m"
	ColorPurple Color = "\033[35m"
	ColorCyan   Color = "\033[36m"
	ColorWhite  Color = "\033[37m"
	ColorBold   Color = "\033[1m"
)

// Config holds progress bar configuration
type Config struct {
	Total           int64   // Total work items
	Width           int     // Bar width in characters
	Style           Style   // Visual style
	Description     string  // Description text
	ShowPercentage  bool    // Show percentage
	ShowCount       bool    // Show current/total count
	ShowETA         bool    // Show estimated time remaining
	ShowSpeed       bool    // Show items per second
	ShowElapsedTime bool    // Show elapsed time
	ShowSpinner     bool    // Show animated spinner
	ColorComplete   Color   // Color for completed portion
	ColorIncomplete Color   // Color for incomplete portion
	Writer          io.Writer // Output writer (default: os.Stderr)
}

// ProgressBar represents a progress bar instance
type ProgressBar struct {
	config      Config
	current     int64
	startTime   time.Time
	lastUpdate  time.Time
	spinnerPos  int
	mu          sync.Mutex
	isFinished  bool
	lastRate    float64
	rateSamples []float64
}

// Spinner characters
var spinnerChars = []string{"⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"}

// New creates a new progress bar with default configuration
func New(total int64) *ProgressBar {
	return NewWithConfig(Config{
		Total:           total,
		Width:           40,
		Style:           StyleDefault,
		ShowPercentage:  true,
		ShowCount:       true,
		ShowETA:         true,
		ShowSpeed:       true,
		ShowElapsedTime: false,
		ShowSpinner:     false,
		Writer:          os.Stderr,
	})
}

// NewWithConfig creates a new progress bar with custom configuration
func NewWithConfig(config Config) *ProgressBar {
	if config.Width <= 0 {
		config.Width = 40
	}
	if config.Writer == nil {
		config.Writer = os.Stderr
	}
	if config.Total < 0 {
		config.Total = 0
	}

	return &ProgressBar{
		config:      config,
		current:     0,
		startTime:   time.Now(),
		lastUpdate:  time.Now(),
		spinnerPos:  0,
		rateSamples: make([]float64, 0, 10),
	}
}

// Add increments the progress by n
func (pb *ProgressBar) Add(n int64) {
	pb.mu.Lock()
	defer pb.mu.Unlock()

	if pb.isFinished {
		return
	}

	pb.current += n
	if pb.current > pb.config.Total {
		pb.current = pb.config.Total
	}
	if pb.current < 0 {
		pb.current = 0
	}

	pb.render()
}

// Set sets the current progress value
func (pb *ProgressBar) Set(current int64) {
	pb.mu.Lock()
	defer pb.mu.Unlock()

	if pb.isFinished {
		return
	}

	pb.current = current
	if pb.current > pb.config.Total {
		pb.current = pb.config.Total
	}
	if pb.current < 0 {
		pb.current = 0
	}

	pb.render()
}

// Increment increments the progress by 1
func (pb *ProgressBar) Increment() {
	pb.Add(1)
}

// Current returns the current progress value
func (pb *ProgressBar) Current() int64 {
	pb.mu.Lock()
	defer pb.mu.Unlock()
	return pb.current
}

// Percentage returns the current percentage (0-100)
func (pb *ProgressBar) Percentage() float64 {
	pb.mu.Lock()
	defer pb.mu.Unlock()

	if pb.config.Total == 0 {
		return 0
	}
	return float64(pb.current) / float64(pb.config.Total) * 100
}

// ETA returns the estimated time remaining
func (pb *ProgressBar) ETA() time.Duration {
	pb.mu.Lock()
	defer pb.mu.Unlock()

	if pb.current == 0 || pb.current >= pb.config.Total {
		return 0
	}

	elapsed := time.Since(pb.startTime).Seconds()
	if elapsed == 0 {
		return 0
	}

	rate := float64(pb.current) / elapsed
	if rate == 0 {
		return 0
	}

	remaining := float64(pb.config.Total-pb.current) / rate
	return time.Duration(remaining) * time.Second
}

// Elapsed returns the elapsed time
func (pb *ProgressBar) Elapsed() time.Duration {
	pb.mu.Lock()
	defer pb.mu.Unlock()
	return time.Since(pb.startTime)
}

// Speed returns the items per second rate
func (pb *ProgressBar) Speed() float64 {
	pb.mu.Lock()
	defer pb.mu.Unlock()

	elapsed := time.Since(pb.startTime).Seconds()
	if elapsed == 0 {
		return 0
	}
	return float64(pb.current) / elapsed
}

// Finish completes the progress bar
func (pb *ProgressBar) Finish() {
	pb.mu.Lock()
	defer pb.mu.Unlock()

	if pb.isFinished {
		return
	}

	pb.current = pb.config.Total
	pb.isFinished = true
	pb.render()
	fmt.Fprintln(pb.config.Writer)
}

// Reset resets the progress bar
func (pb *ProgressBar) Reset() {
	pb.mu.Lock()
	defer pb.mu.Unlock()

	pb.current = 0
	pb.startTime = time.Now()
	pb.lastUpdate = time.Now()
	pb.isFinished = false
	pb.spinnerPos = 0
	pb.rateSamples = make([]float64, 0, 10)
}

// Clear clears the progress bar line
func (pb *ProgressBar) Clear() {
	pb.mu.Lock()
	defer pb.mu.Unlock()

	fmt.Fprint(pb.config.Writer, "\r\033[K")
}

// Describe sets or updates the description
func (pb *ProgressBar) Describe(desc string) {
	pb.mu.Lock()
	defer pb.mu.Unlock()

	pb.config.Description = desc
	pb.render()
}

// UpdateConfig updates the configuration
func (pb *ProgressBar) UpdateConfig(config Config) {
	pb.mu.Lock()
	defer pb.mu.Unlock()

	pb.config = config
	if pb.config.Width <= 0 {
		pb.config.Width = 40
	}
	if pb.config.Writer == nil {
		pb.config.Writer = os.Stderr
	}
	pb.render()
}

// render outputs the progress bar
func (pb *ProgressBar) render() {
	if pb.config.Writer == nil {
		return
	}

	var parts []string

	// Spinner
	if pb.config.ShowSpinner {
		parts = append(parts, spinnerChars[pb.spinnerPos]+" ")
		pb.spinnerPos = (pb.spinnerPos + 1) % len(spinnerChars)
	}

	// Description
	if pb.config.Description != "" {
		parts = append(parts, pb.config.Description+": ")
	}

	// Progress bar
	bar := pb.buildBar()
	parts = append(parts, bar)

	// Percentage (calculate directly to avoid deadlock)
	if pb.config.ShowPercentage {
		var percent float64
		if pb.config.Total > 0 {
			percent = float64(pb.current) / float64(pb.config.Total) * 100
		}
		parts = append(parts, fmt.Sprintf(" %6.2f%%", percent))
	}

	// Count
	if pb.config.ShowCount {
		parts = append(parts, fmt.Sprintf(" %d/%d", pb.current, pb.config.Total))
	}

	// Speed (calculate directly to avoid deadlock)
	if pb.config.ShowSpeed && pb.current > 0 {
		elapsed := time.Since(pb.startTime).Seconds()
		if elapsed > 0 {
			speed := float64(pb.current) / elapsed
			parts = append(parts, fmt.Sprintf(" [%s/s]", formatSpeed(speed)))
		}
	}

	// ETA (calculate directly to avoid deadlock)
	if pb.config.ShowETA && pb.current > 0 && pb.current < pb.config.Total {
		elapsed := time.Since(pb.startTime).Seconds()
		if elapsed > 0 {
			rate := float64(pb.current) / elapsed
			if rate > 0 {
				remaining := float64(pb.config.Total-pb.current) / rate
				eta := time.Duration(remaining) * time.Second
				parts = append(parts, fmt.Sprintf(" ETA: %s", formatDuration(eta)))
			}
		}
	}

	// Elapsed time (calculate directly to avoid deadlock)
	if pb.config.ShowElapsedTime {
		elapsed := time.Since(pb.startTime)
		parts = append(parts, fmt.Sprintf(" [%s]", formatDuration(elapsed)))
	}

	// Build final line
	line := strings.Join(parts, "")

	// Clear line and print
	fmt.Fprint(pb.config.Writer, "\r\033[K"+line)
}

// buildBar constructs the progress bar string
func (pb *ProgressBar) buildBar() string {
	style := pb.config.Style
	width := pb.config.Width

	var completeLen int
	if pb.config.Total > 0 {
		completeLen = int(float64(width) * float64(pb.current) / float64(pb.config.Total))
	}
	if completeLen > width {
		completeLen = width
	}
	if completeLen < 0 {
		completeLen = 0
	}

	incompleteLen := width - completeLen

	// Build bar
	complete := strings.Repeat(style.Complete, completeLen)
	incomplete := strings.Repeat(style.Incomplete, incompleteLen)

	// Apply colors
	if pb.config.ColorComplete != "" {
		complete = string(pb.config.ColorComplete) + complete + string(ColorReset)
	}
	if pb.config.ColorIncomplete != "" {
		incomplete = string(pb.config.ColorIncomplete) + incomplete + string(ColorReset)
	}

	return style.LeftEnd + complete + incomplete + style.RightEnd
}

// formatDuration formats a duration for display
func formatDuration(d time.Duration) string {
	if d < 0 {
		return "--:--:--"
	}

	hours := int(d.Hours())
	minutes := int(d.Minutes()) % 60
	seconds := int(d.Seconds()) % 60

	if hours > 0 {
		return fmt.Sprintf("%d:%02d:%02d", hours, minutes, seconds)
	}
	return fmt.Sprintf("%02d:%02d", minutes, seconds)
}

// formatSpeed formats items per second for display
func formatSpeed(speed float64) string {
	if speed >= 1e9 {
		return fmt.Sprintf("%.1fG", speed/1e9)
	} else if speed >= 1e6 {
		return fmt.Sprintf("%.1fM", speed/1e6)
	} else if speed >= 1e3 {
		return fmt.Sprintf("%.1fK", speed/1e3)
	}
	return fmt.Sprintf("%.0f", speed)
}

// ==================== Static Progress Bars ====================

// Static creates a static progress bar string without animation
func Static(current, total int64, width int, style Style) string {
	return StaticWithConfig(current, total, Config{
		Width: width,
		Style: style,
	})
}

// StaticWithConfig creates a static progress bar with full configuration
func StaticWithConfig(current, total int64, config Config) string {
	if config.Width <= 0 {
		config.Width = 40
	}

	var completeLen int
	if total > 0 {
		completeLen = int(float64(config.Width) * float64(current) / float64(total))
	}
	if completeLen > config.Width {
		completeLen = config.Width
	}
	if completeLen < 0 {
		completeLen = 0
	}

	incompleteLen := config.Width - completeLen

	complete := strings.Repeat(config.Style.Complete, completeLen)
	incomplete := strings.Repeat(config.Style.Incomplete, incompleteLen)

	if config.ColorComplete != "" {
		complete = string(config.ColorComplete) + complete + string(ColorReset)
	}
	if config.ColorIncomplete != "" {
		incomplete = string(config.ColorIncomplete) + incomplete + string(ColorReset)
	}

	return config.Style.LeftEnd + complete + incomplete + config.Style.RightEnd
}

// StaticPercentage creates a progress bar with percentage text
func StaticPercentage(current, total int64, width int, style Style) string {
	bar := Static(current, total, width, style)
	percent := float64(current) / float64(total) * 100
	return fmt.Sprintf("%s %.1f%%", bar, percent)
}

// StaticFull creates a full progress bar with all info
func StaticFull(current, total int64, width int, style Style, desc string) string {
	bar := Static(current, total, width, style)
	percent := float64(current) / float64(total) * 100
	return fmt.Sprintf("%s: %s %.1f%% (%d/%d)", desc, bar, percent, current, total)
}

// ==================== MultiBar Support ====================

// MultiBar manages multiple progress bars simultaneously
type MultiBar struct {
	bars    []*ProgressBar
	prefixes []string
	mu      sync.Mutex
	writer  io.Writer
}

// NewMultiBar creates a new multi-bar manager
func NewMultiBar() *MultiBar {
	return &MultiBar{
		bars:    make([]*ProgressBar, 0),
		prefixes: make([]string, 0),
		writer:  os.Stderr,
	}
}

// AddBar adds a new progress bar to the manager
func (mb *MultiBar) AddBar(total int64, prefix string) *ProgressBar {
	mb.mu.Lock()
	defer mb.mu.Unlock()

	config := Config{
		Total:           total,
		Width:           30,
		Style:           StyleDefault,
		ShowPercentage:  true,
		ShowCount:       false,
		ShowETA:         false,
		ShowSpeed:       false,
		ShowElapsedTime: false,
		Writer:          io.Discard, // Don't write individually
	}

	pb := NewWithConfig(config)
	mb.bars = append(mb.bars, pb)
	mb.prefixes = append(mb.prefixes, prefix)

	return pb
}

// Render renders all progress bars
func (mb *MultiBar) Render() {
	mb.mu.Lock()
	defer mb.mu.Unlock()

	// Move cursor up for each bar
	for i := 0; i < len(mb.bars); i++ {
		fmt.Fprint(mb.writer, "\033[A\033[K")
	}

	// Render each bar
	for i, pb := range mb.bars {
		bar := pb.buildBar()
		percent := pb.Percentage()
		line := fmt.Sprintf("%s %s %.1f%% (%d/%d)\n", mb.prefixes[i], bar, percent, pb.current, pb.config.Total)
		fmt.Fprint(mb.writer, line)
	}
}

// ==================== Spinner ====================

// Spinner represents an animated spinner
type Spinner struct {
	frames   []string
	interval time.Duration
	current  int
	writer   io.Writer
	prefix   string
	suffix   string
	mu       sync.Mutex
	stopCh   chan struct{}
}

// NewSpinner creates a new spinner
func NewSpinner(prefix string) *Spinner {
	return &Spinner{
		frames:   spinnerChars,
		interval: 100 * time.Millisecond,
		current:  0,
		writer:   os.Stderr,
		prefix:   prefix,
		suffix:   "",
		stopCh:   nil,
	}
}

// SetFrames sets custom spinner frames
func (s *Spinner) SetFrames(frames []string) *Spinner {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.frames = frames
	return s
}

// SetInterval sets the animation interval
func (s *Spinner) SetInterval(d time.Duration) *Spinner {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.interval = d
	return s
}

// SetSuffix sets the suffix text
func (s *Spinner) SetSuffix(suffix string) *Spinner {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.suffix = suffix
	return s
}

// Start starts the spinner
func (s *Spinner) Start() *Spinner {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.stopCh != nil {
		return s
	}

	s.stopCh = make(chan struct{})
	go func() {
		ticker := time.NewTicker(s.interval)
		defer ticker.Stop()

		for {
			select {
			case <-s.stopCh:
				return
			case <-ticker.C:
				s.render()
			}
		}
	}()

	return s
}

// Stop stops the spinner
func (s *Spinner) Stop() {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.stopCh != nil {
		close(s.stopCh)
		s.stopCh = nil
		fmt.Fprint(s.writer, "\r\033[K")
	}
}

// Update updates the suffix text
func (s *Spinner) Update(suffix string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.suffix = suffix
}

func (s *Spinner) render() {
	s.mu.Lock()
	defer s.mu.Unlock()

	frame := s.frames[s.current]
	s.current = (s.current + 1) % len(s.frames)
	fmt.Fprintf(s.writer, "\r\033[K%s %s %s", s.prefix, frame, s.suffix)
}

// ==================== Utility Functions ====================

// FormatBytes formats bytes into human-readable string
func FormatBytes(bytes int64) string {
	const (
		KB = 1024
		MB = KB * 1024
		GB = MB * 1024
		TB = GB * 1024
	)

	switch {
	case bytes >= TB:
		return fmt.Sprintf("%.2f TB", float64(bytes)/TB)
	case bytes >= GB:
		return fmt.Sprintf("%.2f GB", float64(bytes)/GB)
	case bytes >= MB:
		return fmt.Sprintf("%.2f MB", float64(bytes)/MB)
	case bytes >= KB:
		return fmt.Sprintf("%.2f KB", float64(bytes)/KB)
	default:
		return fmt.Sprintf("%d B", bytes)
	}
}

// FormatNumber formats a number with thousands separator
func FormatNumber(n int64) string {
	s := fmt.Sprintf("%d", n)
	result := ""
	for i, c := range s {
		if i > 0 && (len(s)-i)%3 == 0 {
			result += ","
		}
		result += string(c)
	}
	return result
}

// CalculateProgress calculates progress percentage safely
func CalculateProgress(current, total int64) float64 {
	if total == 0 {
		return 0
	}
	percent := float64(current) / float64(total) * 100
	if percent > 100 {
		return 100
	}
	if percent < 0 {
		return 0
	}
	return percent
}

// EstimateTime estimates remaining time based on progress
func EstimateTime(elapsed time.Duration, current, total int64) time.Duration {
	if current == 0 || current >= total {
		return 0
	}

	rate := float64(current) / elapsed.Seconds()
	if rate == 0 {
		return 0
	}

	remaining := float64(total-current) / rate
	return time.Duration(remaining) * time.Second
}

// ==================== Iterator Support ====================

// Iterate creates a progress bar for iterating over a slice
func Iterate[T any](items []T, desc string, fn func(int, T) error) error {
	pb := New(int64(len(items)))
	pb.config.Description = desc
	pb.config.Width = 30

	for i, item := range items {
		if err := fn(i, item); err != nil {
			return err
		}
		pb.Increment()
	}

	pb.Finish()
	return nil
}

// IterateWithConfig creates a progress bar with config for iterating
func IterateWithConfig[T any](items []T, config Config, fn func(int, T) error) error {
	config.Total = int64(len(items))
	pb := NewWithConfig(config)

	for i, item := range items {
		if err := fn(i, item); err != nil {
			return err
		}
		pb.Increment()
	}

	pb.Finish()
	return nil
}